pkg/util/timeutil/ptp/ptp_clock_linux_test.go

```diff
@@ -28,3 +31,29 @@ func TestClockNow(t *testing.T) {
 		t.Errorf("clock mismatch: got %v; timeutil says %v", got, want)
 	}
 }
+
+func TestInitClock(t *testing.T) {
+	var file *os.File
+	var err error
+	var devOpenerCallCount int
+	clock := &Clock{
+		devOpener: func(name string) (*os.File, error) {
+			require.Equal(t, "/dev/ptp-test", name)
+			file, err = os.CreateTemp("", "")
+			require.NoError(t, err)
+			require.NotNil(t, file)
+			devOpenerCallCount++
+			return file, err
+		},
```

Comment:
It seems like the only benefit of mocking this func is to check that a file is being opened. Also, the fact that we're calling this `devOpener` does not guarantee that the "real" file/device will be opened, because the code is non-trivial, there is a branch between the test-only and non-test code. Finally, I'm not sure that `/dev/ptp0` is a "regular" file like `CreateTemp` creates, and that we're comparing apples to apples.

I think it's low benefit, perhaps it's ok to remove `devOpener` and only mock `cGetTime`.

Is there a way to put some data into the tmp file so that `C.clock_gettime` would read it as time? If so, I think there would be more benefit in doing this instead of having `devOpener`, i.e.:
- At the beginning of the test, create a real tmp file, and write a few timestamps to it (in the spec-compatible format).
- Have the test read time from this file.

This way we would:
- have more confidence in the fact that the "real" code opens the file/device
- not need `devOpener` and `cGetTime` mocks altogether
- have no performance impact from branching/mocking

Reply:
It isn't possible to write timestamps in a plain file and make the code read them. I don't see another way to go through the code that opens a clock character device in unit test. Also the `TestClockNow` that you added is sufficient to test the `gettime` call without mocking. So the question is - Is there anything besides what is already there that could be done to improve the coverage? If there isn't - I'll just close the PR.

Reply:
Yeah, perhaps it's not trivial to write such a device. Ideally, that's what we would need to do for a more end-to-end test.

Agree that the existing test already covers the unit-testable bits. As I said above, I don't see benefit in mocking `devOpener`, but maybe mocking `C.clock_gettime`/`cGetTime` has some value in checking that the conversion between C time and Go time works. This is sort of tested by `TestClockNow` though, so maybe the extra branching that we're adding here is not worth it.

The only downside of `TestClockNow` is that it's non-deterministic, and depends on system time. A mocked `cGetTime` test  would be more precise.

Reply:
We should also have a proper end-to-end test running with a vMotion PTP device anyway, which would cover this.

Reply:
Ok, I'm going to close the PR now. I just few tests however and I think it isn't that hard to test with a real device. This works for example on my gce worker:
```
darin@gceworker-darin$ uname -a
Linux gceworker-darin 6.4.11-060411-generic #202308161732 SMP PREEMPT_DYNAMIC Wed Aug 16 17:41:09 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

darin@gceworker-darin$ modprobe ptp_kvm

darin@gceworker-darin$ ls -la /dev/ptp0
crw------- 1 root root 246, 0 Aug 18 15:29 /dev/ptp0

darin@gceworker-darin$ sudo ./testptp -g -d /dev/ptp0
clock time: 1692377627.723922889 or Fri Aug 18 16:53:47 2023
```

The ptp_kvm module makes the host clock available in the VM similarly to how VMware does it. Unfortunately this module is not compiled with the flavor of linux that we use for our teamcity agents. You should ask the test-eng or dev-inf to update the files in /build/packer to use newer version and to load the module at startup. With that it should be possible to run integration tests on a real device.